okta/okta.go

```diff
@@ -0,0 +1,321 @@
+type CreationOptions struct {
+	ClientID     string
+	ClientSecret string
+	RedirectURI  string
+	Issuer       string
+	CookiePath   string
+	CookieDomain string
+	CookieTTL    time.Duration
+}
+
+func (opts CreationOptions) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(opts.ClientID == "", "must specify client ID")
+	catcher.NewWhen(opts.ClientSecret == "", "must specify client secret")
+	catcher.NewWhen(opts.RedirectURI == "", "must specify redirect URI")
+	catcher.NewWhen(opts.Issuer == "", "must specify issuer")
+	catcher.NewWhen(opts.CookiePath == "", "must specify cookie path")
+	catcher.NewWhen(opts.CookieDomain == "", "must specify cookie domain")
+	if opts.CookieTTL == time.Duration(0) {
+		opts.CookieTTL = time.Hour
+	}
+	return catcher.Resolve()
+}
+
+type userManager struct {
+	clientID     string
+	clientSecret string
+	redirectURI  string
+	issuer       string
+	cookiePath   string
+	cookieDomain string
+	cookieTTL    time.Duration
+
+	// TODO (kim): token caching functions
+}
+
+// NewUserManager creates a manager that connects to Okta for user
+// management services.
+func NewUserManager(opts CreationOptions) (gimlet.UserManager, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid Okta manager options")
+	}
+	m := &userManager{
+		clientID:     opts.ClientID,
+		clientSecret: opts.ClientSecret,
+		redirectURI:  opts.RedirectURI,
+		issuer:       opts.Issuer,
+		cookiePath:   opts.CookiePath,
+		cookieDomain: opts.CookieDomain,
+		cookieTTL:    opts.CookieTTL,
+	}
+	return m, nil
+}
+
+func (m *userManager) GetUserByToken(ctx context.Context, token string) (gimlet.User, error) {
+	return nil, errors.New("not implemented")
+}
+
+func (m *userManager) CreateUserToken(user string, password string) (string, error) {
+	return "", errors.New("not implemented")
+}
+
+const (
+	nonceCookieName = "okta-nonce"
+	stateCookieName = "okta-state"
+)
+
+func (m *userManager) GetLoginHandler(callbackURL string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// TODO (kim): persist nonce and state.
+		nonce, err := randomString()
+		if err != nil {
+			grip.Error(message.WrapError(err, message.Fields{
+				"message": "could not get login handler",
+			}))
+			gimlet.WriteResponse(w, gimlet.MakeTextErrorResponder(errors.Wrap(err, "could not get login handler")))
+			return
+		}
+		state, err := randomString()
+		if err != nil {
+			grip.Error(message.WrapError(err, message.Fields{
+				"message": "could not get login handler",
+			}))
+			gimlet.WriteResponse(w, gimlet.MakeTextErrorResponder(errors.Wrap(err, "could not get login handler")))
+			return
+		}
+
+		q := r.URL.Query()
+		q.Add("client_id", m.clientID)
+		q.Add("response_type", "code")
+		q.Add("response_mode", "query")
+		q.Add("scope", "openid")
+		q.Add("redirect_uri", m.redirectURI)
+		q.Add("state", state)
+		q.Add("nonce", nonce)
+
+		http.SetCookie(w, &http.Cookie{
```

Comment:
Are you sure you want to reset this every time the function is called? (I have no idea how often it's called)

Reply:
The state and nonce should be unique to the request they're created in. If the user has to authenticate again for some reason, they have to present a different state/nonce.

I was thinking of doing something similar to what we do with LDAP right now. With LDAP, we only bind if the user is logged out. Reauthorization occurs once an hour or so and just does a simple check on the user's groups. Ideally, this login function should be called pretty infrequently since I would like the authentication to only occur if they aren't logged into Okta or after some long TTL has expired on the login cookie (right now I think the TTL is a year).